pkg/importer/http_data_source.go

```diff
@@ -0,0 +1,271 @@
+	if dp.cancel != nil {
+		dp.cancel()
+	}
+	return err
+}
+
+func createHTTPClient(certDir string) (*http.Client, error) {
+	client := &http.Client{
+		// Don't set timeout here, since that will be an absolute timeout, we need a relative to last progress timeout.
+	}
+
+	if certDir == "" {
+		return client, nil
+	}
+
+	// let's get system certs as well
+	certPool, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, errors.Wrap(err, "Error getting system certs")
+	}
+
+	files, err := ioutil.ReadDir(certDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error listing files in %s", certDir)
+	}
+
+	for _, file := range files {
+		if file.IsDir() || file.Name()[0] == '.' {
+			continue
+		}
+
+		fp := path.Join(certDir, file.Name())
+
+		klog.Infof("Attempting to get certs from %s", fp)
+
+		certs, err := ioutil.ReadFile(fp)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error reading file %s", fp)
+		}
+
+		if ok := certPool.AppendCertsFromPEM(certs); !ok {
+			klog.Warningf("No certs in %s", fp)
+		}
+	}
+
+	client.Transport = &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: certPool,
+		},
+	}
+
+	return client, nil
+}
+
+func createHTTPReader(ctx context.Context, ep *url.URL, accessKey, secKey, certDir string) (io.ReadCloser, error) {
+	client, err := createHTTPClient(certDir)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating http client")
+	}
+
+	client.CheckRedirect = func(r *http.Request, via []*http.Request) error {
+		if len(accessKey) > 0 && len(secKey) > 0 {
+			r.SetBasicAuth(accessKey, secKey) // Redirects will lose basic auth, so reset them manually
+		}
+		return nil
+	}
+
+	// http.NewRequest can only return error on invalid METHOD, or invalid url. Here the METHOD is always GET, and the url is always valid, thus error cannot happen.
+	req, _ := http.NewRequest("GET", ep.String(), nil)
+
+	req = req.WithContext(ctx)
+	if len(accessKey) > 0 && len(secKey) > 0 {
+		req.SetBasicAuth(accessKey, secKey)
+	}
+	klog.V(2).Infof("Attempting to get object %q via http client\n", ep.String())
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "HTTP request errored")
+	}
+	if resp.StatusCode != 200 {
+		klog.Errorf("http: expected status code 200, got %d", resp.StatusCode)
+		return nil, errors.Errorf("expected status code 200, got %d. Status: %s", resp.StatusCode, resp.Status)
+	}
+	countingReader := &util.CountingReader{
+		Reader:  resp.Body,
+		Current: 0,
+	}
+	return countingReader, nil
+}
+
+func (dp *HTTPDataSource) pollProgress(reader *util.CountingReader, idleTime, pollInterval time.Duration) {
+	count := reader.Current
+	lastUpdate := time.Now()
+	for {
+		if count < reader.Current {
+			// Some progress was made, reset now.
+			lastUpdate = time.Now()
+			count = reader.Current
+		}
+
+		if time.Until(lastUpdate.Add(idleTime)).Nanoseconds() < 0 {
+			// No progress for the idle time, cancel http client.
+			dp.cancel() // This will trigger dp.ctx.Done()
```

Comment:
What happens we reach this line and dp.cancel() is called? Shall dp.cancel be set to nil in order  for dp.cancel() not to be called second time on provider.Close()?

Can here be a race between Polling thread (that reaches this line) and main thread that runs provider.Close()?

Reply:
I think you are right, there is a potential race between the two. I also should set dp.cancel to nil so the other can't be called and vice versa, when I call the other one I should set it to nil there as well. I will add a lock to this code so there won't be a problem.